{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "def two_break_dist(P, Q):\n",
    "\n",
    "    graph = defaultdict(list)\n",
    "\n",
    "    for perm_cycle in P + Q :\n",
    "\n",
    "        L = len(perm_cycle)\n",
    "\n",
    "        for i in xrange(len(perm_cycle)) :\n",
    "            graph[perm_cycle[i]].append(-1 * perm_cycle[(i + 1) % L])\n",
    "            graph[-1 * perm_cycle[(i + 1) % L]].append(perm_cycle[i])\n",
    "\n",
    "    component_count = 0\n",
    "    remaining = set(graph.keys())\n",
    "\n",
    "    while len(remaining) > 0 :\n",
    "        component_count += 1\n",
    "        queue = [remaining.pop()]\n",
    "\n",
    "        while queue :\n",
    "            current = queue.pop(0)\n",
    "            queue += filter(lambda node: node in remaining, graph.get(current, []))\n",
    "            remaining -= set(queue)\n",
    "\n",
    "    return sum(map(len,P)) - component_count\n",
    "\n",
    "input_file = open(\"untitled.txt\", 'r')\n",
    "P, Q = [line.strip().lstrip('(').rstrip(')').split(')(') for line in input_file.readlines()]\n",
    "P = [map(int, block.split()) for block in P]\n",
    "Q = [map(int, block.split()) for block in Q]\n",
    "\n",
    "dist = two_break_dist(P, Q)\n",
    "\n",
    "print str(dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '+4)(+5'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-16-4c2b25612f92>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0minput_file\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"untitled.txt\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'r'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m \u001b[0mp\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmap\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mint\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minput_file\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'\\n'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m''\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m \u001b[1;33m:\u001b[0m \u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m' '\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      6\u001b[0m \u001b[0mq\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmap\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mint\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minput_file\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'\\n'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m''\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m \u001b[1;33m:\u001b[0m \u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m' '\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: invalid literal for int() with base 10: '+4)(+5'"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import random\n",
    "\n",
    "input_file = open(\"untitled.txt\", 'r')\n",
    "p = map(int, input_file.readline().replace('\\n', '')[1 : -1].split(' '))\n",
    "q = map(int, input_file.readline().replace('\\n', '')[1 : -1].split(' '))\n",
    "\n",
    "def ChromosomeToCycle(chromosome) :\n",
    "    \n",
    "    nodes = []\n",
    "    \n",
    "    for i in chromosome :\n",
    "        if(i > 0) :\n",
    "            nodes.append(2 * i - 1)\n",
    "            nodes.append(2 * i)\n",
    "        else :\n",
    "            nodes.append(-2 * i)\n",
    "            nodes.append(-2 * i - 1)\n",
    "            \n",
    "    return nodes\n",
    "\n",
    "def CycleToChromosome(nodes) :\n",
    "    \n",
    "    chromosome = []\n",
    "    \n",
    "    for j in range(1, len(nodes) / 2 + 1) :\n",
    "        if nodes[2 * j - 2] < nodes[2 * j - 1] : chromosome.append(nodes[2 * j - 1]/2)\n",
    "        else : chromosome.append(-nodes[2 * j - 2] / 2)\n",
    "    return chromosome\n",
    "\n",
    "def ColoredEdges(p) :\n",
    "    \n",
    "    edges = []\n",
    "    \n",
    "    if(type(p[0]) == int) :\n",
    "        chromosome = p\n",
    "        chromosome.append(chromosome[0])\n",
    "        nodes = ChromosomeToCycle(chromosome)\n",
    "        for j in range(1, len(chromosome)) : edges.append([nodes[2 * j - 1],nodes[2 * j]])\n",
    "        chromosome.pop()\n",
    "        \n",
    "        return edges\n",
    "    \n",
    "    else :\n",
    "        for chromosome in p :\n",
    "            temp = chromosome[0]\n",
    "            chromosome.append(temp)\n",
    "            nodes = ChromosomeToCycle(chromosome)\n",
    "            for j in range(1, len(chromosome)) : edges.append([nodes[2 * j - 1],nodes[2 * j]])\n",
    "            chromosome.pop()\n",
    "            \n",
    "        return edges\n",
    "    \n",
    "def get_blacked_edges(col_edges) :\n",
    "    \n",
    "    blacked_edges = []\n",
    "    \n",
    "    for node in col_edges :\n",
    "        temp = []\n",
    "        for num in node :\n",
    "            if(num % 2 == 0) : temp.append(num / 2)\n",
    "            else : temp.append(-(num + 1) / 2)\n",
    "        blacked_edges.append(temp)\n",
    "        \n",
    "    return blacked_edges\n",
    "\n",
    "def GraphToGenome(col_edges) :\n",
    "    \n",
    "    blacked_edges = get_blacked_edges(col_edges)\n",
    "    chromosome_list = []\n",
    "    \n",
    "    while blacked_edges != [] :\n",
    "        \n",
    "        start, end = blacked_edges[0]\n",
    "        blacked_edges.remove(blacked_edges[0])\n",
    "        \n",
    "        loop_end = -1 * start\n",
    "        chromosome = [start]\n",
    "        \n",
    "        while end != loop_end :\n",
    "            for i in range(len(blacked_edges)) :\n",
    "                if(blacked_edges[i][0] == -1 * end) :\n",
    "                    start,end = blacked_edges[i]\n",
    "                    blacked_edges.remove(blacked_edges[i])\n",
    "                    break\n",
    "                    \n",
    "                elif(blacked_edges[i][1] == -1 * end) :\n",
    "                    blacked_edges[i] = blacked_edges[i][::-1] \n",
    "                    start,end = blacked_edges[i]\n",
    "                    blacked_edges.remove(blacked_edges[i])\n",
    "                    break\n",
    "        \n",
    "            chromosome.append(start)\n",
    "        chromosome_list.append(chromosome)\n",
    "        \n",
    "    return chromosome_list\n",
    "\n",
    "def two_breakongenomegraph(col_edges, i, j, k, l) :\n",
    "    \n",
    "    if [i,j] in col_edges : col_edges.remove([i, j])\n",
    "    else : col_edges.remove([j,i])\n",
    "        \n",
    "    if [k,l] in col_edges : col_edges.remove([k, l])\n",
    "    else : col_edges.remove([l, k])\n",
    "        \n",
    "    col_edges.append([i, k])\n",
    "    col_edges.append([j, l])\n",
    "    \n",
    "    return col_edges\n",
    "\n",
    "def two_breakongenome(p, i, j, k, l) :\n",
    "    \n",
    "    colored_edges = ColoredEdges(p)\n",
    "    genomegraph = two_breakongenomegraph(colored_edges, i, j, k, l)\n",
    "    ans = GraphToGenome(genomegraph)\n",
    "    \n",
    "    return ans\n",
    "\n",
    "def shortestrearrangement(p,q) :\n",
    "\n",
    "    print '(' + ' '.join('+' + str(i) if i > 0 else str(i) for i in p) + ')'\n",
    "    \n",
    "    red_edges = ColoredEdges(p)\n",
    "    blue_edges = ColoredEdges(q)\n",
    "    \n",
    "    breakpointgraph = red_edges + blue_edges\n",
    "    \n",
    "    while True :\n",
    "        \n",
    "        red_edges = ColoredEdges(p)\n",
    "        \n",
    "        for i in range(len(red_edges)) :\n",
    "            if red_edges[i][::-1] in blue_edges : red_edges[i] = red_edges[i][::-1]\n",
    "        if(sorted(red_edges) == sorted(blue_edges)) : break\n",
    "            \n",
    "        temp_list = []\n",
    "        \n",
    "        for blue_edge in blue_edges :\n",
    "            if blue_edge not in red_edges : temp_list.append(blue_edge)\n",
    "                \n",
    "        node = random.choice(temp_list)\n",
    "        \n",
    "        for red_edge in red_edges :\n",
    "            if node[0] in red_edge : to_edge = red_edge\n",
    "            elif node[1] in red_edge : from_edge = red_edge\n",
    "                \n",
    "        if((to_edge[1]==node[0]) & (from_edge[0]==node[1])) :\n",
    "            red_edges.remove(from_edge)\n",
    "            red_edges.remove(to_edge)\n",
    "            red_edges.append([to_edge[1],from_edge[0]])\n",
    "            red_edges.append([to_edge[0],from_edge[1]])\n",
    "\n",
    "        elif((to_edge[0]==node[0]) & (from_edge[0]==node[1])) :\n",
    "            red_edges.remove(from_edge)\n",
    "            red_edges.remove(to_edge)\n",
    "            to_edge = to_edge[::-1]\n",
    "            red_edges.append([to_edge[1],from_edge[0]])\n",
    "            red_edges.append([to_edge[0],from_edge[1]])\n",
    "    \n",
    "        elif((to_edge[1]==node[0]) & (from_edge[1]==node[1])) :\n",
    "            red_edges.remove(from_edge)\n",
    "            red_edges.remove(to_edge)\n",
    "            from_edge = from_edge[::-1]\n",
    "            red_edges.append([to_edge[1],from_edge[0]])\n",
    "            red_edges.append([to_edge[0],from_edge[1]])\n",
    "            \n",
    "        elif((to_edge[0]==node[0]) & (from_edge[1]==node[1])) :\n",
    "            red_edges.remove(from_edge)\n",
    "            red_edges.remove(to_edge)\n",
    "            from_edge = from_edge[::-1]\n",
    "            to_edge = to_edge[::-1]\n",
    "            red_edges.append([to_edge[1],from_edge[0]])\n",
    "            red_edges.append([to_edge[0],from_edge[1]])\n",
    "            \n",
    "        breakpointgraph = red_edges + blue_edges\n",
    "        p = two_breakongenome(p, to_edge[0], to_edge[1], from_edge[1], from_edge[0])\n",
    "        \n",
    "        str_p = ''\n",
    "        \n",
    "        for res in p : str_p += '(' + ' '.join('+' + str(i) if i > 0 else str(i) for i in res) + ')'\n",
    "        print str_p\n",
    "        \n",
    "        if len(p) == 1 : p = p[0]\n",
    "\n",
    "    return None\n",
    "\n",
    "shortestrearrangement(p, q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '(+1 +2 +3 +4)(+5 -9 -8 -7 +6)'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-4-4d0a8bc32bd3>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0minput_file\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"untitled.txt\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'r'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 6\u001b[1;33m \u001b[0mk\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0minput_file\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'\\n'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m''\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      7\u001b[0m \u001b[0mp\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0minput_file\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'\\n'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m''\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      8\u001b[0m \u001b[0mq\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0minput_file\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'\\n'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m''\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: invalid literal for int() with base 10: '(+1 +2 +3 +4)(+5 -9 -8 -7 +6)'"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "from collections import defaultdict\n",
    "\n",
    "input_file = open(\"untitled.txt\", 'r')\n",
    "\n",
    "k = int(input_file.readline().replace('\\n', ''))\n",
    "p = input_file.readline().replace('\\n', '')\n",
    "q = input_file.readline().replace('\\n', '')\n",
    "\n",
    "def rev_comp(sequence) :\n",
    "\n",
    "    rev_comp = ''\n",
    "    \n",
    "    for nucleotide in sequence :\n",
    "        if(nucleotide == 'A') : rev_comp = rev_comp + 'T'\n",
    "        elif(nucleotide == 'T') : rev_comp = rev_comp + 'A'\n",
    "        elif(nucleotide == 'G') : rev_comp = rev_comp + 'C'\n",
    "        elif(nucleotide == 'C') : rev_comp = rev_comp + 'G'\n",
    "            \n",
    "    rev_comp = rev_comp[::-1]\n",
    "    return rev_comp\n",
    "\n",
    "def sharedkmer(dna1,dna2) :\n",
    "\n",
    "    dna1_dict = defaultdict(list)\n",
    "    for i in xrange(len(dna1) - k + 1) : dna1_dict[dna1[i : i + k]].append(i)\n",
    "        \n",
    "    shared_kmer_indices = set()\n",
    "    \n",
    "    for j in xrange(len(dna2) - k + 1) :\n",
    "        shared_kmer_indices |= set(map(lambda x : (x, j), dna1_dict[dna2[j : j + k]]))\n",
    "        shared_kmer_indices |= set(map(lambda x : (x, j), dna1_dict[rev_comp(dna2[j : j + k])]))\n",
    "        \n",
    "    return shared_kmer_indices\n",
    "\n",
    "ans = sharedkmer(p, q)\n",
    "for res in ans : print res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
